### ISSUE TITLE
请问能提供在已有模型上继续进行指令微调的训练参数吗？万分感谢

### ISSUE THREAD

**Issue by matrixssy** [ymcui/Chinese-LLaMA-Alpaca#114]
我发现在新的GPT4中英文50K数据上继续微调loss很大，基本不收敛了

**Comment by matrixssy** [comment]
是不是prompter的格式变了或者是什么别的原因

**Comment by airaria** [comment]
请描述一下详细情况（学习率、batch size、loss等）

**Comment by matrixssy** [comment]
> 请描述一下详细情况（学习率、batch size、loss等）

`Training Alpaca-LoRA model with params:
base_model: ../models/llama-7b-hf
data_path: ../dataset/alpaca_data_gpt4.json
output_dir: ../models/alpaca-lora-7b-chinese-train
batch_size: 128
micro_batch_size: 4
num_epochs: 1
learning_rate: 0.0003
cutoff_len: 512
val_set_size: 2000
lora_r: 8
lora_alpha: 32
lora_dropout: 0.1
lora_target_modules: ['q_proj', 'v_proj', 'k_proj', 'o_proj', 'gate_proj', 'down_proj', 'up_proj']
train_on_inputs: True
group_by_length: False
wandb_project: 
wandb_run_name: 
wandb_watch: 
wandb_log_model: 
resume_from_checkpoint: False
prompt template: alpaca

Loading checkpoint shards: 100%|??????????????????????????????????????????????????????????????????????????????????????????????????????| 33/33 [00:15<00:00,  2.07it/s]
Found cached dataset json (/home/ifpp/.cache/huggingface/datasets/json/default-b00d1a8c7fb9fc29/0.0.0/fe5dd6ea2639a6df622901539cb550cf8797e5a6b2dd7af1cf934bed8e233e6e)
100%|??????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????| 1/1 [00:00<00:00, 317.65it/s]
trainable params: 19988480 || all params: 6905483264 || trainable%: 0.28945809056123434
Loading cached split indices for dataset at /home/ifpp/.cache/huggingface/datasets/json/default-b00d1a8c7fb9fc29/0.0.0/fe5dd6ea2639a6df622901539cb550cf8797e5a6b2dd7af1cf934bed8e233e6e/cache-42c3775f0d21c0a6.arrow and /home/ifpp/.cache/huggingface/datasets/json/default-b00d1a8c7fb9fc29/0.0.0/fe5dd6ea2639a6df622901539cb550cf8797e5a6b2dd7af1cf934bed8e233e6e/cache-c8b37057e973382c.arrow
{'loss': 282.9425, 'learning_rate': 2.9999999999999997e-06, 'epoch': 0.03}                                                                                            
  3%|????                                                                                                                        | 10/390 [29:59<19:08:02, 181.27s/it`

**Comment by matrixssy** [comment]
> 请描述一下详细情况（学习率、batch size、loss等）

lr=3e-4, bs=128, loss > 100
所有配置都与原始的alpaca lora类似，唯一不确定的是是否需要对 'gate_proj', 'down_proj', 'up_proj'这三个层进行低秩分解（但通过尝试loss均>100）

**Comment by matrixssy** [comment]
以下是我的训练代码，我是用的是经过合并过后的hf模型参数（已经在推理上验证其参数加载正确）
`import os
import sys
from typing import List

import fire
import torch
import transformers
from peft import PeftModel
from datasets import load_dataset

"""
Unused imports:
import torch.nn as nn
import bitsandbytes as bnb
"""

from peft import (
    LoraConfig,
    get_peft_model,
    get_peft_model_state_dict,
    prepare_model_for_int8_training,
    set_peft_model_state_dict,
)
from transformers import LlamaForCausalLM, LlamaTokenizer

from utils.prompter import Prompter


def train(
    # model/data params
    base_model: str = '../models/alpaca-lora-7b-chinese-merged',  # the only required argument
    data_path: str = '../dataset/GPT4/alpaca_gpt4_data_zh.json',
    output_dir: str = "../models/alpaca-lora-7b-chinese-train",
    # training hyperparams
    batch_size: int = 128,
    micro_batch_size: int = 4,
    num_epochs: int = 1,
    learning_rate: float = 3e-4,
    cutoff_len: int = 256,
    val_set_size: int = 2000,
    # lora hyperparams
    lora_r: int = 8,
    lora_alpha: int = 32,
    lora_dropout: float = 0.1,
    lora_target_modules: List[str] = [
        "q_proj",
        "v_proj",
        "k_proj",
        "o_proj"
    ],
    # llm hyperparams
    train_on_inputs: bool = True,  # if False, masks out inputs in loss
    group_by_length: bool = False,  # faster, but produces an odd training loss curve
    # wandb params
    wandb_project: str = "",
    wandb_run_name: str = "",
    wandb_watch: str = "",  # options: false | gradients | all
    wandb_log_model: str = "",  # options: false | true
    resume_from_checkpoint: str = None,  # either training checkpoint or final adapter
    prompt_template_name: str = "alpaca",  # The prompt template to use, will default to alpaca.
):
    if int(os.environ.get("LOCAL_RANK", 0)) == 0:
        print(
            f"Training Alpaca-LoRA model with params:\n"
            f"base_model: {base_model}\n"
            f"data_path: {data_path}\n"
            f"output_dir: {output_dir}\n"
            f"batch_size: {batch_size}\n"
            f"micro_batch_size: {micro_batch_size}\n"
            f"num_epochs: {num_epochs}\n"
            f"learning_rate: {learning_rate}\n"
            f"cutoff_len: {cutoff_len}\n"
            f"val_set_size: {val_set_size}\n"
            f"lora_r: {lora_r}\n"
            f"lora_alpha: {lora_alpha}\n"
            f"lora_dropout: {lora_dropout}\n"
            f"lora_target_modules: {lora_target_modules}\n"
            f"train_on_inputs: {train_on_inputs}\n"
            f"group_by_length: {group_by_length}\n"
            f"wandb_project: {wandb_project}\n"
            f"wandb_run_name: {wandb_run_name}\n"
            f"wandb_watch: {wandb_watch}\n"
            f"wandb_log_model: {wandb_log_model}\n"
            f"resume_from_checkpoint: {resume_from_checkpoint or False}\n"
            f"prompt template: {prompt_template_name}\n"
        )
    assert (
        base_model
    ), "Please specify a --base_model, e.g. --base_model='decapoda-research/llama-7b-hf'"
    gradient_accumulation_steps = batch_size // micro_batch_size

    prompter = Prompter(prompt_template_name)

    device_map = "auto"
    world_size = int(os.environ.get("WORLD_SIZE", 1))
    ddp = world_size != 1
    if ddp:
        device_map = {"": int(os.environ.get("LOCAL_RANK") or 0)}
        gradient_accumulation_steps = gradient_accumulation_steps // world_size

    # Check if parameter passed or if set within environ
    use_wandb = len(wandb_project) > 0 or (
        "WANDB_PROJECT" in os.environ and len(os.environ["WANDB_PROJECT"]) > 0
    )
    # Only overwrite environ if wandb param passed
    if len(wandb_project) > 0:
        os.environ["WANDB_PROJECT"] = wandb_project
    if len(wandb_watch) > 0:
        os.environ["WANDB_WATCH"] = wandb_watch
    if len(wandb_log_model) > 0:
        os.environ["WANDB_LOG_MODEL"] = wandb_log_model

    model = LlamaForCausalLM.from_pretrained(
        base_model,
        load_in_8bit=True,
        torch_dtype=torch.float16,
        device_map=device_map,
    )

    tokenizer = LlamaTokenizer.from_pretrained(base_model)
    tokenizer.save_pretrained(output_dir)
    print(f"Extended vocabulary size: {len(tokenizer)}")

    tokenizer.pad_token_id = (
        0  # unk. we want this to be different from the eos token
    )
    tokenizer.padding_side = "left"  # Allow batched inference

    model = prepare_model_for_int8_training(model)

    config = LoraConfig(
        r=lora_r,
        lora_alpha=lora_alpha,
        target_modules=lora_target_modules,
        lora_dropout=lora_dropout,
        inference_mode=False,
        bias="none",
        task_type="CAUSAL_LM",
    )

    model = get_peft_model(model, config)
    model.print_trainable_parameters()  # Be more transparent about the % of trainable params.

    def tokenize(prompt, add_eos_token=True):
        # there's probably a way to do this with the tokenizer settings
        # but again, gotta move fast
        result = tokenizer(
            prompt,
            truncation=True,
            max_length=cutoff_len,
            padding=False,
            return_tensors=None,
        )
        if (
            result["input_ids"][-1] != tokenizer.eos_token_id
            and len(result["input_ids"]) < cutoff_len
            and add_eos_token
        ):
            result["input_ids"].append(tokenizer.eos_token_id)
            result["attention_mask"].append(1)

        result["labels"] = result["input_ids"].copy()

        return result

    def generate_and_tokenize_prompt(data_point):
        full_prompt = prompter.generate_prompt(
            data_point["instruction"],
            data_point["input"],
            data_point["output"],
        )
        tokenized_full_prompt = tokenize(full_prompt)
        if not train_on_inputs:
            user_prompt = prompter.generate_prompt(
                data_point["instruction"], data_point["input"]
            )
            tokenized_user_prompt = tokenize(user_prompt, add_eos_token=False)
            user_prompt_len = len(tokenized_user_prompt["input_ids"])

            tokenized_full_prompt["labels"] = [
                -100
            ] * user_prompt_len + tokenized_full_prompt["labels"][
                user_prompt_len:
            ]  # could be sped up, probably
        return tokenized_full_prompt

    if data_path.endswith(".json") or data_path.endswith(".jsonl"):
        data = load_dataset("json", data_files=data_path)
    else:
        data = load_dataset(data_path)

    if resume_from_checkpoint:
        # Check the available weights and load them
        checkpoint_name = os.path.join(
            resume_from_checkpoint, "pytorch_model.bin"
        )  # Full checkpoint
        if not os.path.exists(checkpoint_name):
            checkpoint_name = os.path.join(
                resume_from_checkpoint, "adapter_model.bin"
            )  # only LoRA model - LoRA config above has to fit
            resume_from_checkpoint = (
                False  # So the trainer won't try loading its state
            )
        # The two files above have a different name depending on how they were saved, but are actually the same.
        if os.path.exists(checkpoint_name):
            print(f"Restarting from {checkpoint_name}")
            adapters_weights = torch.load(checkpoint_name)
            set_peft_model_state_dict(model, adapters_weights)
        else:
            print(f"Checkpoint {checkpoint_name} not found")

    if val_set_size > 0:
        train_val = data["train"].train_test_split(
            test_size=val_set_size, shuffle=True, seed=42
        )
        train_data = (
            train_val["train"].shuffle().map(generate_and_tokenize_prompt)
        )
        val_data = (
            train_val["test"].shuffle().map(generate_and_tokenize_prompt)
        )
    else:
        train_data = data["train"].shuffle().map(generate_and_tokenize_prompt)
        val_data = None

    if not ddp and torch.cuda.device_count() > 1:
        # keeps Trainer from trying its own DataParallelism when more than 1 gpu is available
        model.is_parallelizable = True
        model.model_parallel = True

    trainer = transformers.Trainer(
        model=model,
        train_dataset=train_data,
        eval_dataset=val_data,
        args=transformers.TrainingArguments(
            per_device_train_batch_size=micro_batch_size,
            gradient_accumulation_steps=gradient_accumulation_steps,
            warmup_steps=100,
            num_train_epochs=num_epochs,
            learning_rate=learning_rate,
            fp16=True,
            logging_steps=1,
            optim="adamw_torch",
            evaluation_strategy="steps" if val_set_size > 0 else "no",
            save_strategy="steps",
            eval_steps=10 if val_set_size > 0 else None,
            save_steps=10,
            output_dir=output_dir,
            save_total_limit=3,
            load_best_model_at_end=True if val_set_size > 0 else False,
            ddp_find_unused_parameters=False if ddp else None,
            group_by_length=group_by_length,
            report_to="wandb" if use_wandb else "none",
            run_name=wandb_run_name if use_wandb else "none",
        ),
        data_collator=transformers.DataCollatorForSeq2Seq(
            tokenizer, pad_to_multiple_of=8, return_tensors="pt", padding=True
        ),
    )
    model.config.use_cache = False

    old_state_dict = model.state_dict
    model.state_dict = (
        lambda self, *_, **__: get_peft_model_state_dict(
            self, old_state_dict()
        )
    ).__get__(model, type(model))

    if torch.__version__ >= "2" and sys.platform != "win32":
        model = torch.compile(model)

    trainer.train(resume_from_checkpoint=resume_from_checkpoint)

    model.save_pretrained(output_dir)

    print(
        "\n If there's a warning about missing keys above, please disregard :)"
    )


if __name__ == "__main__":
    fire.Fire(train)
`

**Comment by airaria** [comment]
可否print一下train_data[0]?

以及
```
tokenizer.pad_token_id = (
    0  # unk. we want this to be different from the eos token
)
tokenizer.padding_side = "left"  # Allow batched inference
```
这处删掉试试

**Comment by matrixssy** [comment]
> 

是采用默认alpaca.json template生成的输入数据，万分感谢抽空回答，我生成后在下边评论~

**Comment by matrixssy** [comment]
> 可否print一下train_data[0]?
> 
> 以及
> 
> ```
> tokenizer.pad_token_id = (
>     0  # unk. we want this to be different from the eos token
> )
> tokenizer.padding_side = "left"  # Allow batched inference
> ```
> 
> 这处删掉试试

`{'instruction': 'Read the following article and come up with two discussion questions.', 'input': "In today's society, the amount of technology usage by children has grown dramatically over the last decade. This increased usage has led to both good and bad outcomes, with many parents wondering how they should navigate their children's device usage.", 'output': "1. What are some potential positive and negative consequences of increased technology usage by children?\n2. In what ways can parents monitor their children's device usage to ensure that it is not harmful to their development?", 'input_ids': [1, 13866, 338, 385, 15278, 393, 16612, 263, 3414, 29892, 3300, 2859, 411, 385, 1881, 393, 8128, 4340, 3030, 29889, 14350, 263, 2933, 393, 7128, 2486, 1614, 2167, 278, 2009, 29889, 13, 13, 2277, 29937, 2799, 4080, 29901, 13, 6359, 278, 1494, 564, 41925, 280, 322, 2041, 701, 411, 1023, 2313, 375, 40972, 5155, 29889, 13, 13, 2277, 29937, 10567, 29901, 13, 797, 9826, 29915, 29879, 12459, 29892, 278, 5253, 310, 15483, 8744, 491, 4344, 756, 21633, 8541, 19574, 975, 278, 1833, 316, 6332, 29889, 910, 11664, 8744, 756, 5331, 304, 1716, 1781, 322, 4319, 714, 26807, 29892, 411, 1784, 11825, 9873, 920, 896, 881, 23624, 1009, 4344, 29915, 29879, 4742, 8744, 29889, 13, 13, 2277, 29937, 13291, 29901, 13, 32902, 1724, 526, 777, 7037, 6374, 322, 8178, 27721, 310, 11664, 15483, 8744, 491, 4344, 29973, 13, 33378, 512, 825, 5837, 508, 11825, 11819, 1009, 4344, 29915, 29879, 4742, 8744, 304, 9801, 393, 372, 338, 451, 10311, 1319, 304, 1009, 5849, 29973, 2], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], 'labels': [1, 13866, 338, 385, 15278, 393, 16612, 263, 3414, 29892, 3300, 2859, 411, 385, 1881, 393, 8128, 4340, 3030, 29889, 14350, 263, 2933, 393, 7128, 2486, 1614, 2167, 278, 2009, 29889, 13, 13, 2277, 29937, 2799, 4080, 29901, 13, 6359, 278, 1494, 564, 41925, 280, 322, 2041, 701, 411, 1023, 2313, 375, 40972, 5155, 29889, 13, 13, 2277, 29937, 10567, 29901, 13, 797, 9826, 29915, 29879, 12459, 29892, 278, 5253, 310, 15483, 8744, 491, 4344, 756, 21633, 8541, 19574, 975, 278, 1833, 316, 6332, 29889, 910, 11664, 8744, 756, 5331, 304, 1716, 1781, 322, 4319, 714, 26807, 29892, 411, 1784, 11825, 9873, 920, 896, 881, 23624, 1009, 4344, 29915, 29879, 4742, 8744, 29889, 13, 13, 2277, 29937, 13291, 29901, 13, 32902, 1724, 526, 777, 7037, 6374, 322, 8178, 27721, 310, 11664, 15483, 8744, 491, 4344, 29973, 13, 33378, 512, 825, 5837, 508, 11825, 11819, 1009, 4344, 29915, 29879, 4742, 8744, 304, 9801, 393, 372, 338, 451, 10311, 1319, 304, 1009, 5849, 29973, 2]}`

**Comment by airaria** [comment]
prompt不完全一致，Chinese-alpaca在精调时用了统一的模版，不再区分带不带input，见 #7

**Comment by airaria** [comment]
在训练时只会计算"response:"后面的损失，但你的labels是全都计算了。这块可以参加原版Stanford Alpaca的实现。

**Comment by matrixssy** [comment]
> 可否print一下train_data[0]?
> 
> 以及
> 
> ```
> tokenizer.pad_token_id = (
>     0  # unk. we want this to be different from the eos token
> )
> tokenizer.padding_side = "left"  # Allow batched inference
> ```
> 
> 这处删掉试试

删掉后就loss就正常了~

**Comment by matrixssy** [comment]
> 在精调时只会计算"response:"后面的损失，但从你的labels看，是整个序列全都计算了。这块可以参考原版Stanford Alpaca的实现。

好的，我也去尝试以下只计算"response:"后面的损失

**Comment by matrixssy** [comment]
已解决！

**Comment by magicleo** [comment]
@matrixssy 能提供下最新的代码学习下吗？

**Comment by ylinlinz** [comment]
如果删掉 tokenizer.pad_token_id = (
    0  # unk. we want this to be different from the eos token
)这行，就会报错没有 pad_token
